### ISSUE TITLE
False negative for incorrect syntax in functions

### ISSUE THREAD

**Issue by malliaridis** [anboralabs/intellij-firebase-highlighter#36]
There are two syntax errors that are not recognized by the syntax highlighter. The first is the starting `&&` which is in some cases correctly recognized and in others not, and the invalid continuation of the statement after a semicolon (see image below).

![Screenshot_2](https://user-images.githubusercontent.com/33470669/104358066-40343b00-550e-11eb-9c6c-705cef51690f.jpg)

**Comment by dalgarins** [comment]
Which plugin version was used for this issue?
 
I'm trying to reproduce with version 2.1.0 and I wasn't able to reproduce the issue.

**Comment by malliaridis** [comment]
I updated to the latest version too (2.1.0-SNAPSHOT) and I think there is a more general problem related with this.

It seems like the whole autocomplete does not work and when I try to delete for example a line break with backspace it won't let me do so.
Invalidating and restarting IDE and also reinstalling plugin did not help. I get the following error from the plugin when starting IDE Webstorm:

```
com.intellij.diagnostic.PluginException: Incorrect name specified in <fileType>, should be Firebase Rules, actual Firebase Rule File [Plugin: co.anbora.labs.firebase-syntax-highlighting]
	at com.intellij.openapi.fileTypes.impl.FileTypeManagerImpl.instantiateFileTypeBean(FileTypeManagerImpl.java:427)
	at com.intellij.openapi.fileTypes.impl.FileTypeManagerImpl.mergeOrInstantiateFileTypeBean(FileTypeManagerImpl.java:384)
	at com.intellij.openapi.fileTypes.impl.FileTypeManagerImpl.instantiatePendingFileTypes(FileTypeManagerImpl.java:376)
	at com.intellij.openapi.fileTypes.impl.FileTypeManagerImpl.getRegisteredFileTypes(FileTypeManagerImpl.java:730)
	at com.intellij.psi.stubs.StubUpdatingIndex.instantiateElementTypesFromFields(StubUpdatingIndex.java:577)
	at com.intellij.psi.stubs.StubUpdatingIndex.lambda$ensureSerializationManagerInitialized$0(StubUpdatingIndex.java:569)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:658)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:610)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:65)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeNonCancelableSection(CoreProgressManager.java:218)
	at com.intellij.psi.stubs.StubUpdatingIndex.ensureSerializationManagerInitialized(StubUpdatingIndex.java:568)
	at com.intellij.psi.stubs.StubUpdatingIndex.getValueExternalizer(StubUpdatingIndex.java:336)
	at com.intellij.util.indexing.FileBasedIndexImpl.createIndexStorage(FileBasedIndexImpl.java:484)
	at com.intellij.util.indexing.FileBasedIndexImpl.initIndexStorage(FileBasedIndexImpl.java:431)
	at com.intellij.util.indexing.FileBasedIndexImpl.registerIndexer(FileBasedIndexImpl.java:398)
	at com.intellij.util.indexing.FileBasedIndexDataInitialization.lambda$initAssociatedDataForExtensions$0(FileBasedIndexDataInitialization.java:65)
	at com.intellij.util.indexing.IndexInfrastructure$DataInitialization.executeNestedInitializationTask(IndexInfrastructure.java:174)
	at com.intellij.util.indexing.IndexInfrastructure$DataInitialization.runParallelNestedInitializationTasks(IndexInfrastructure.java:162)
	at com.intellij.util.indexing.IndexInfrastructure$DataInitialization.call(IndexInfrastructure.java:123)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:216)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$200(BoundedTaskExecutor.java:27)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.execute(BoundedTaskExecutor.java:195)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:208)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.run(BoundedTaskExecutor.java:184)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

IDE Info:
```
WebStorm 2020.3.1
Build #WS-203.6682.155, built on December 28, 2020
[...]
Runtime version: 11.0.9.1+11-b1145.63 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 10 10.0
GC: ParNew, ConcurrentMarkSweep
Memory: 1974M
Cores: 16
Non-Bundled Plugins: co.anbora.labs.firebase-syntax-highlighting
```

**Comment by malliaridis** [comment]
And some sample code (I reduced it to the minimum and left some functions undefined):
```
rules_version = '2';

// Rulesp
service cloud.firestore {
  match /databases/{database}/documents {

    match /{prefix=**}/{document} {
        allow read, write: if isAdmin();
    }

    match /articles/{articleId} {

        function isCorrectArticle() {
            let article = get(/databases/$(database)/documents/articles/$(articleId)).data;
            return unchangedKeys.hasAll(REQUIRED_ARTICLE_KEYS())
                   && unchangedKeys.hasOnly(REQUIRED_ARTICLE_KEYS())
                   && addedKeys.size() == 0;
        }

        allow read: if isSignedIn();

        allow create: if isAuthor();

        allow update: if isModerator();

        allow delete: if false;
    }
  }
}
```
Result:
![Screenshot_1](https://user-images.githubusercontent.com/33470669/104595185-84405080-5672-11eb-84be-a2510b61efd0.jpg)
Error text in `let`: `RETURN_KEYWORD expected, got 'let'`. 

The error mentioned in #35 also invalidates the above error.

**Comment by dalgarins** [comment]
the first issue related with error in the IDE was fixed in version: 2.1.1, also the comment related with && conditions was fixed.

about the 'let' keyword, I'm going to work on that. I'm going to create another issue for that.

**Comment by malliaridis** [comment]
Sounds great, thanks for the quick response and fix to the issue. :D
